src/karapace/compatibility/schema_compatibility.py

```diff
@@ -0,0 +1,145 @@
+"""
+Copyright (c) 2024 Aiven Ltd
+See LICENSE for details
+"""
+from avro.compatibility import (
+    merge,
+    ReaderWriterCompatibilityChecker as AvroChecker,
+    SchemaCompatibilityResult,
+    SchemaCompatibilityType,
+    SchemaIncompatibilityType,
+)
+from avro.schema import Schema as AvroSchema
+from jsonschema import Draft7Validator
+from karapace.compatibility import CompatibilityModes
+from karapace.compatibility.jsonschema.checks import compatibility as jsonschema_compatibility, incompatible_schema
+from karapace.compatibility.protobuf.checks import check_protobuf_schema_compatibility
+from karapace.protobuf.schema import ProtobufSchema
+from karapace.schema_models import ParsedTypedSchema, ValidatedTypedSchema
+from karapace.schema_type import SchemaType
+from karapace.utils import assert_never
+
+import logging
+
+LOG = logging.getLogger(__name__)
+
+
+class SchemaCompatibility:
+    @staticmethod
+    def check_compatibility(
+        old_schema: ParsedTypedSchema,
+        new_schema: ValidatedTypedSchema,
+        compatibility_mode: CompatibilityModes,
+    ) -> SchemaCompatibilityResult:
+        """Check that `old_schema` and `new_schema` are compatible under `compatibility_mode`."""
+
+        if compatibility_mode is CompatibilityModes.NONE:
+            LOG.info("Compatibility level set to NONE, no schema compatibility checks performed")
+            return SchemaCompatibilityResult(SchemaCompatibilityType.compatible)
+
+        if old_schema.schema_type is not new_schema.schema_type:
+            return incompatible_schema(
+                incompat_type=SchemaIncompatibilityType.type_mismatch,
+                message=f"Comparing different schema types: {old_schema.schema_type} with {new_schema.schema_type}",
+                location=[],
+            )
+
+        if old_schema.schema_type is SchemaType.AVRO:
+            assert isinstance(old_schema.schema, AvroSchema)
+            assert isinstance(new_schema.schema, AvroSchema)
+            if compatibility_mode in {CompatibilityModes.BACKWARD, CompatibilityModes.BACKWARD_TRANSITIVE}:
+                result = SchemaCompatibility.check_avro_compatibility(
+                    reader_schema=new_schema.schema,
+                    writer_schema=old_schema.schema,
+                )
+
```

Comment:
nit: why those empty lines between the `if` and `elif`?

Reply:
Not sure, I think it was pre-existing code that I moved around.
I have now removed the empty lines so it's more compact :)